src/System.Net.NameResolution/src/System/Net/DNS.cs

```diff
@@ -0,0 +1,383 @@
+                this.address = address;
+            }
+
+            internal readonly string hostName;
+            internal bool includeIPv6;
+            internal IPAddress address;
+        }
+
+        private static void ResolveCallback(object context)
+        {
+            ResolveAsyncResult result = (ResolveAsyncResult)context;
+            IPHostEntry hostEntry;
+            try
+            {
+                if (result.address != null)
+                {
+                    hostEntry = InternalGetHostByAddress(result.address, result.includeIPv6);
+                }
+                else
+                {
+                    hostEntry = InternalGetHostByName(result.hostName, result.includeIPv6);
+                }
+            }
+            catch (Exception exception)
+            {
+                if (exception is OutOfMemoryException)
+                    throw;
+
+                result.InvokeCallback(exception);
+                return;
+            }
+
+            result.InvokeCallback(hostEntry);
+        }
+
+        // Helpers for async GetHostByName, ResolveToAddresses, and Resolve - they're almost identical
+        // If hostName is an IPString and justReturnParsedIP==true then no reverse lookup will be attempted, but the orriginal address is returned.
+        private static IAsyncResult HostResolutionBeginHelper(string hostName, bool justReturnParsedIp, AsyncCallback requestCallback, object state)
+        {
+            if (hostName == null)
+            {
+                throw new ArgumentNullException("hostName");
+            }
+
+            GlobalLog.Print("Dns.HostResolutionBeginHelper: " + hostName);
+
+            // See if it's an IP Address.
+            IPAddress address;
+            ResolveAsyncResult asyncResult;
+            if (IPAddress.TryParse(hostName, out address))
+            {
+                if ((address.Equals(IPAddress.Any) || address.Equals(IPAddress.IPv6Any)))
+                {
+                    throw new ArgumentException(SR.net_invalid_ip_addr, "hostNameOrAddress");
+                }
+
+                asyncResult = new ResolveAsyncResult(address, null, true, state, requestCallback);
+
+                if (justReturnParsedIp)
+                {
+                    IPHostEntry hostEntry = NameResolutionUtilities.GetUnresolvedAnswer(address);
+                    asyncResult.StartPostingAsyncOp(false);
+                    asyncResult.InvokeCallback(hostEntry);
+                    asyncResult.FinishPostingAsyncOp();
+                    return asyncResult;
+                }
+            }
+            else
+            {
+                asyncResult = new ResolveAsyncResult(hostName, null, true, state, requestCallback);
+            }
+
+            // Set up the context, possibly flow.
+            asyncResult.StartPostingAsyncOp(false);
+
+            // Start the resolve.
+            Task.Factory.StartNew(ResolveCallback, asyncResult);
```

Comment:
This would be better as:
```C#
Task.Factory.StartNew(s => ResolveCallback(s), asyncResult,
    CancellationToken.None, TaskCreationOptions.DenyChildAttach, TaskScheduler.Default);
```
That does two things:
1. It allows the compiler to cache the delegate.  Unfortunately today the compiler won't cache delegates for method groups like this.
2. It ensures that the task is scheduled on to the default scheduler (using the same options as for Task.Run) vs using whatever scheduler is current.